onnxruntime/core/framework/graph_partitioner.cc

```diff
@@ -52,7 +54,190 @@ static void BuildFusedKernelDef(KernelDefBuilder& builder, const IndexedSubGraph
+  std::unordered_set<std::string_view> layout_sensitive_ops{
+      "Resize", "Conv", "QLinearConv", "FusedConv", "AveragePool", "QLinearAveragePool", "GlobalAveragePool",
+      "QLinearGlobalAveragePool", "MaxPool", "GlobalMaxPool", "LRN"};
+
+  for (std::unique_ptr<onnx_layout_transformation::api::NodeRef>& node : api_graph->Nodes()) {
+    if (layout_sensitive_ops.count(node->OpType())) {
+      if (node->GetExecutionProviderType() != current_ep.Type()) {
+        continue;
+      }
+
+      auto domain = node->Domain();
+      // Skip if domain is incorrect
+      if (domain != kOnnxDomain && domain != kOnnxDomainAlias && domain != kMSDomain) {
+        continue;
+      }
+
+      // Skip if already transformed
+      if (node->GetAttributeIntDefault("channels_last", 0) == 1) {
+        continue;
+      }
+
+      // Skip if unknown rank
+      auto shape = NodeFromApiNode(*node).InputDefs()[0]->Shape();
+      if (shape == nullptr) {
+        continue;
+      }
+
+      // Convert to channels last
+      size_t rank = shape->dim_size();
+
+      bool has_channel_last_attr = node->GetAttributeInt("channels_last") != nullopt ? true : false;
+      if (has_channel_last_attr) {
+        node->SetAttributeInt("channels_last", 1);
+      }
+
+      auto input_perm = onnx_layout_transformation::ChannelFirstToLastPerm(rank);
+      auto output_perm = onnx_layout_transformation::ChannelLastToFirstPerm(rank);
+
+      // Resize is special case. Older versions of resize have a bug where ROI and Scales
+      // cannot be made empty inputs. To handle this case we need to jump a few extra hoops
+      // to make sure its inputs are correctly handled.
+      if (node->OpType() == "Resize") {
+        // Current code skips handling ROI.
+        // ROI needs special handling. Enable it when an EP which supports ROI starts using 
+        // layout transformer. NNAPI which currently used layout transformer does not support it.
+        std::vector<const std::vector<int64_t>*> input_perms{&input_perm, nullptr};
+        for (size_t i = 2; i < node->Inputs().size(); i++) {
+          auto constant = api_graph->GetConstant(node->Inputs()[i]);
+          if (constant != nullptr && constant->Data().size() > 0) {
+            input_perms.push_back(&input_perm);
+          } else {
+            input_perms.push_back(nullptr);
+          }
+        }
+        onnx_layout_transformation::WrapTransposesAroundNode(*api_graph, *node, input_perms, {&output_perm});
+      } else {
+        onnx_layout_transformation::WrapTransposesAroundNode(*api_graph, *node, {&input_perm}, {&output_perm});
+      }
+
+      if (!has_channel_last_attr) {
+        onnx_layout_transformation::SwapNodeOpTypeAndDomain(*api_graph, *node, node->OpType(), kMSNHWCDomain);
+      }
+      modified = true;
+    }
+  }
+
+  modified = onnx_layout_transformation::Optimize(*api_graph,
+                                                  /*allow_extended_ops*/ true,
```

Comment:
we need to allow extended ops because like you said NNAPI supports a subset of these ops. 
The layout transformer only transforms the layout for ops which are claimed by nnapi ... so if there exists a QLinearConv op which is not assigned to NNAPI then we do not transform it.

The transpose optimizer will do a similar thing - it will not push the transpose across any layout sensitive op which is not assigned to the provided EP. 

So in case where we have say following nodes 
Transpose (NHWC-> NCHW)
QuantizeLinear
QLinearConv
QLinearAveragePool
DequantizeLinear
Transpose(NCHW -> NHWC)

If the entire model is picked by NNAPI then the graph will look like this
QuantizeLinear
QLinearConv
QLinearAveragePool
DequantizeLinear

And if NNAPI does not pick QLinearConv then it will look like this:
QuantizeLinear
Transpose(NHWC->NCHW)
QLinearConv
Transpose(NCHW -> NHWC)
QLinearAveragePool
DequantizeLinear

I will add this test